juju/juju
# Do not set link-layer data via SetInstanceInfo

MAAS was the only provider setting link-layer data via `SetInstanceInfo`, also making it the only provider using the old `SetParentLinkLayerDevicesBeforeTheirChildren` method.

This was causing issues whereby provider IDs were inserted into the global collection, but the devices were set without provider IDs (see deleted code in this patch for specific TODOs regarding this).

In turn the instance-poller was unable to set these IDs, because seeing them in the global collection indicated that they were already in use.

This patch stops `SetInstanceInfo` from setting link-layer data and removes `SetParentLinkLayerDevicesBeforeTheirChildren`. Tests using this method now use `SetLinkLayerDevices` directly. That method is deprecated and has had its validation logic and associated tests removed to ease testing - it is only used in tests.

Work to follow will prevent MAAS from sending link-layer data with `StartInstanceResult` and will probably remove the members from the type completely.

## QA steps

- Bootstrap to MAAS and deploy something (tested on MAAS 2.3 and 2.8).
- Check that there are no logs about not assigning provider IDs to NICs.
- Check that the `linklayerdevices` collection shows documents with provider IDs.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1897330
